### DISCUSSION TITLE
Checkout Pro: It's painfully hard to test

### DISCUSSION THREAD

**Discussion by rafaelfaria** [mercadopago/sdk-js#62] [General]
I am not sure what is going on, but today I am trying to run some tests and it works intermittently, meaning 1 in 10 attempts would go through.

I go through all the screens all the way to payment, I do EXACTLY the same thing over and over again, the most frustrating thing is that the only error I get is

`Ops, ocorreu um erro.`

![Screen Shot 2022-04-22 at 11 55 28 pm](https://user-images.githubusercontent.com/227917/164728741-6974c3e1-4682-4fa4-b109-2c6c9aa748d8.png)

Seriously, how is this helpful?!

What error? How do we get help with this kind of message?

I noticed that some stuff has changed today, so I am attributing this behaviour to whatever changes you guys made today.

If this helps, you can try to debug from this.

I am using these cards for testing: https://www.mercadopago.com.br/developers/pt/docs/checkout-pro/additional-content/test-cards

And i am getting this sort of URL
```
https://www.mercadopago.com.br/checkout/v1/payment/redirect/04aa34f8-abe6-4bd7-abef-d38bab5782e7/fatal/?preference-id=5390060-afdd837a-876d-4fb8-9a5e-82a5585e1183&router-request-id=1beec3db-6e44-402b-899b-21ce7eed53c7&p=fd9ba87ab9251325a94f8e054725f466
```

I am almost giving up on integrating with Mercado Pago.

**Comment by eltinMeli** [comment]
Hi @rafaelfaria

We apologize for the inconvenience in your workflow.
Even though this space is open, it is not intended to be a tool to support unexpected errors of the platform and its integrations.

**We have an official channel prepared to solve your case quickly.** 

So if the error still persists, please pile up all the information and access the following link to share the problem with us: https://www.mercadopago.com/developers/pt/support/center/contact

Thank you so much for your contact!

**Reply by rafaelfaria** [reply]
Thanks.

**Comment by tomasmalio** [comment]
Hi @rafaelfaria you solved the problem because I continue with the same error than you. I'm working with a marketplace, sellers and buyers but when they press at the end of the process that page appears.

**Comment by carlosoro** [comment]
Its horrible how much pain the MP sandbox does to a product testing flow when this happens. Right now we are unable to test some features in a easy way because of this error. The support center doesnt solve anything , they just make random recommendations and keep passing the ball between different support people

**Reply by tomasmalio** [reply]
It's something that I can understand WHY they do this!

You're testing the flow online in production with real users and real money?

**Reply by carlosoro** [reply]
In our case, production is still working. But we are used to this kind of scenarios where we cant test our payment flow for a couple of days because they broke the API/SDK. It's great to know that I'm not alone, but they don't seem to care anymore. I'm really curious about how its their test and release workflow, because it's really common that something important gets trashed and they left the flow useless

**Reply by tomasmalio** [reply]
I've tested the app in production and everything goes well so the problem is with user tests.

**Reply by masterquiroga** [reply]
We too can't test the MercadoPago Checkout Pro using my test token and either test users or the test cards.

While I know this may not be the channel, I still want to use this as a heads up for MercadoPago (just because this is what a technical person from MercadoPago can most likely see):
Please fix the sandbox for the Checkout Pro; it's just plainly unacceptable that we need to test our integration in production, either from the bad ethical practice of having to pay for the tests, and also from the obviously bad QA practice of testing in production.

I'm saying this as with the intention to help you since I had a lot of faith due to your seemly unbreakable UX (and if I'd have to choose I'd use something from latam which probably helps our local economies lol); just think that if you fix this, your developers won't have to shy away to other platforms such as the good ol' Stripe or PayPal. 

Best regards.

**Comment by vsilva472** [comment]
Today i still getting this error while trying split payments on sandbox mode.  
The attribute `sandbox_init_point` drives me to the same screen of dead.
![image](https://user-images.githubusercontent.com/4265802/196051721-a15a4250-5541-4660-becb-a16e35e7d50b.png)

**Reply by bonino97** [reply]
Same.

**Comment by clovisdanielss** [comment]
The problem is usually because you are trying to make the payment with the same account you'll recieve.

**Comment by amouratoglou** [comment]
same here... its not the same account :/ 
also getting error pages every couple attempts mercadopago turns into mercadolibre WTF.
test users do not work.

**Comment by stanleytakamatsu** [comment]
Same here...  
For me, the error occurs when I use the integrated seller's token via OAuth to generate the Checkout Pro.  

I generate both production and test access tokens, but both result in an error in the sandbox checkout, making testing impossible.  

When I generate the checkout using the application's own account credentials, I need to use the production `access_token` to test the sandbox. However, in this case, I am able to complete the payment.

**Comment by amouratoglou** [comment]
I made tests with real payment to be able to test the flow. will give a try to what you mentioned.

**Comment by bauca** [comment]
For those arriving in this post because they are having issues with testing payments:
If you are using **Checkout Pro**:
1. Make sure your return URLs (success,failure,pending) are correctly set, AND using HTTPS.
2. Make sure you are using a test account as the Seller.
3. Make sure you are using a test account as the Buyer.
4. Using the Seller(test) account, retrieve the credentials for the Notification Webhooks AND the Production Credentials (access token).
5. Use both keys retrieved in step (4) in your test environment.
Your application should be able to process the payments correctly.

Any misconfiguration in the aforementioned items will throw you the error: "Ops, ocorreu um erro" (Which I honestly say it is useless from a Developer's point of view.

**Comment by kAutoman** [comment]
I'm testing MercadoPago checkoutpro integration with express.js using test credit card.
<img width="641" height="741" alt="image" src="https://github.com/user-attachments/assets/a5cdac66-2bef-4cd2-ba32-cd08214a6a4b" />

I put all the correct info, but getting this screen.
Already tried all possible cards. but still facing same issue.
<img width="1600" height="654" alt="image" src="https://github.com/user-attachments/assets/8ec0ea84-a69d-47f7-a86f-8e1fb5e09182" />
Anyone have experience with this?

**Reply by lucasBruma** [reply]
I'm experiencing the same issue! @kAutoman, have you tried using another card? I've tried all of them https://www.mercadopago.com.ar/developers/es/docs/your-integrations/test/cards, but none of them work.

**Reply by Benjamin-Urzua-Mar** [reply]
Hi, it's been a while from this post, but maybe you were choosing the wrong checkout integration (Pro, Bricks, API). I was having a similar problem and the solution was change that. Hope it helps ;)

**Comment by chnt-gomez** [comment]
I'm also facing issues with CheckoutPro. Right now, after using the sandbox init point with test cards, I'm getting a very weird CORS error:

Access to fetch at 'https://api.mercadopago.com/v1/card_tokens?public_key=<REDACTED>&locale=es-MX&referer=https%3A%2F%2Fsandbox.mercadopago.com.mx%2Fcheckout%2Fv1%2Fredirect%2Fb89e9ef3-f847-44b0-a9a7-7f5bb2d3168a%2Fcard-form%2F%3Fpreference-id%3D3293893257-cfa203e1-5862-4516-8cb4-4a4eb1eb225d%26p%3Dc272c9c90ce0469f9698b35a90d01093' from origin 'https://secure-fields.mercadopago.com' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource.

Im dropping the integration and moving forward with another provider

**Reply by lemazzo** [reply]
Hi, we apologize for the error. It has now been fixed, and the tests are working again.
Thank you for reporting it.

**Reply by rjslegall** [reply]
Erro Mercado Pago Checkout Transparente

Plataforma: Adobe Commerce / Magento 2.4.8-p4, PHP 8.4, módulo oficial MercadoPago_AdbPayment em modo Sandbox/Teste.

Contexto: ontem o fluxo Sandbox funcionou corretamente: pedido criado com método mercadopago_adbpayment_cc, pagamento aprovado/acreditado. Hoje, ao tentar finalizar novo pedido de teste, a tokenização do cartão falhou antes de chegar ao Magento.

Erro principal no navegador:

Access to fetch at 'https://api.mercadopago.com/v1/card_tokens?...'
from origin 'https://secure-fields.mercadopago.com'
has been blocked by CORS policy:
No 'Access-Control-Allow-Origin' header is present.

Alguém está tendo esse problema?

**Reply by prendonzxor** [reply]
Hey I have the same issue blocked by CORS Policy

**Reply by costanzd-molinos** [reply]
Hello! I'm having the same issue.

**Reply by rjslegall** [reply]
O problema ainda persiste. Ainda não conseguimos resolver, nosso projeto está parado. Já usamos diversas APIs: Paypal, Stripe, Cielo, Inter... de longe essa é a mais complexa e difícil de testar...

**Reply by rjslegall** [reply]
Para o checkout não travar em looping quando o erro acima acontece fizemos esse ajuste no trecho abaixo do mp-sdk.js:


        /**
         * Display Error in Field
         * @param {Array|Object|String} error
         * @return {void}
         */
        displayErrorInField(error) {
            var previousField = undefined;
            var errors = [];

            if (Array.isArray(error)) {
                errors = error;
            } else if (error && Array.isArray(error.errorMessages)) {
                errors = error.errorMessages;
            } else if (error) {
                errors = [error];
            }

            if (!errors.length) {
                return;
            }

            errors.forEach((item) => {
                var field;
                var msg;
                var fieldsMage;

                if (!item) {
                    return;
                }

                field = item.field || item.name || null;
                msg = item.message || item.error || item.cause || item;

                if (!field || previousField === field) {
                    previousField = field;
                    return;
                }

                fieldsMage = {
                    cardNumber: this.fields.fieldCcNumber,
                    securityCode: this.fields.fieldSecurityCode,
                    expirationMonth: this.fields.fieldExpMonth,
                    expirationYear: this.fields.fieldExpYear,
                };

                if (fieldsMage[field]) {
                    validateFormSF.singleToogleValidityState(
                        fieldsMage[field],
                        this.getMessageError(String(msg))
                    );
                }

                previousField = field;
            });
        },
